### ISSUE TITLE
Multiple edges with same name and same source, destination pairs

### ISSUE THREAD

**Issue by parteekcoder** [ControlCore-Project/concore-editor#90]
Are Multiple edges with same names are allowed?

![Screenshot_20230220_203905](https://user-images.githubusercontent.com/100110395/220142947-f4a85bf8-f508-44d6-aec9-9e4650b4488c.png)

**Comment by pradeeban** [comment]
Are Multiple edges with same names are allowed?
Yes, when that is a hyperedge. An edge that connects the same source with multiple destinations.

You won't be able to make an edge with the same name (CU) from XZ to PZ, though, as that is not a hyperedge.

DHGWorkflow (the parent of concore editor) is a directed hypergraph editor. https://github.com/NISYSLAB/DHGWorkflow/

It's a feature!

In concore, we allow connecting two nodes with multiple edges and multiple destination nodes with the same hyperedge from the same source.

But, in your example, you are connecting the same source-destination pair with multiple identical edges. Is that needed? That is more of a research question on the use cases of concore. I keep the question open for a while.

**Comment by parteekcoder** [comment]
> Are Multiple edges with same names are allowed? Yes, when that is a hyperedge. An edge that connects the same source with multiple destinations.
> 
> You won't be able to make an edge with the same name (CU) from XZ to PZ, though, as that is not a hyperedge.
> 
> DHGWorkflow (the parent of concore editor) is a directed hypergraph editor. https://github.com/NISYSLAB/DHGWorkflow/
> 
> It's a feature!
> 
> In concore, we allow connecting two nodes with multiple edges and multiple destination nodes with the same hyperedge from the same source.
> 
> But, in your example, you are connecting the same source-destination pair with multiple identical edges. Is that needed? That is more of a research question on the use cases of concore. I keep the question open for a while.

ok @pradeeban thanks for clarification , In my perspective identical edges between same nodes are not required 

But Once it is confirmed please tell me

**Comment by pradeeban** [comment]
Yes.

In concore, the representation goes:
edge = (source, [destination]).

So, these two cases are actual use cases.
case 1 (a hyperedge):
abc = (A, [B, C])

case 2 (two edges between a pair of nodes):
ab1 = (A, B)
ab2 = (A, B)

But the case you have shown is,
case 3:
ab = (A, [B, B])

I will confirm and either close this question (if the above case 3 is valid) or convert this into a bug (if the case 3 is invalid and should be prevented).

**Comment by parteekcoder** [comment]
hi @pradeeban have you confirmed this Question

**Comment by pradeeban** [comment]
The above scenario as depicted in the screenshot should not exist.
 ab = (A, [B, B])

**Comment by parteekcoder** [comment]
> The above scenario as depicted in the screenshot should not exist. ab = (A, [B, B])

so can you please assign me, I will make an PR shortly to solve this

Thanks @pradeeban for clarification

**Comment by parteekcoder** [comment]
@pradeeban can this be the case

A and B nodes

A to B with edge parteek
B to A with edge parteek

**Comment by pradeeban** [comment]
No, that is a simple loop with the same name for the edges. Not supported in concore. If you try it now, it will throw the message, "Edge with same label exists." (That is the correct behavior).

**Comment by pradeeban** [comment]
Please make sure the PR should not mistakenly open up to allow more cases. We support a subset of hyperedges and avoid the rest.

edge = (source, [destination]) is allowed.
edge = ([source], destination) is not allowed.
edge = ([source], [destination]) is not allowed.
edge = source -> mid-node -> destination is not allowed.

The above cases are implemented correctly.

**Comment by parteekcoder** [comment]
> Please make sure the PR should not mistakenly open up to allow more cases. We support a subset of hyperedges and avoid the rest.
> 
> edge = (source, [destination]) is allowed. edge = ([source], destination) is not allowed. edge = ([source], [destination]) is not allowed. edge = source -> mid-node -> destination is not allowed.
> 
> The above cases are implemented correctly.

OK please give some time , as i was working in issue #92 
I will solve it as early as possible

**Comment by pradeeban** [comment]
No worries. Take your time.

These are good to fix. But not as urgent.

**Comment by parteekcoder** [comment]
> Please make sure the PR should not mistakenly open up to allow more cases. We support a subset of hyperedges and avoid the rest.
> 
> edge = (source, [destination]) is allowed. edge = ([source], destination) is not allowed. edge = ([source], [destination]) is not allowed. edge = source -> mid-node -> destination is not allowed.
> 
> The above cases are implemented correctly.

hey @pradeeban can you please give me some clearity betweeb case 1 and 3

Please tell me if I am right:
Case 1:
There can be  edges with same label from a single source node to multiple nodes
Case 3:
It means there can't be duplicate(same label) edges

How these 2 can occur simultaneously ,correct me if I am wrong

**Comment by pradeeban** [comment]
Case 1 is a typical hyperedge (you can Google it). One edge connecting more than 2 nodes.

Case 3 is just creating identical edges twice with the same label. It is meaningless/duplicate and should not be allowed.

**Comment by pradeeban** [comment]
Your screenshot is case 3 and shouldn't be allowed. (Not a critical issue. You can leave it as is, if in doubt, for now.)
